include/dice/template-library/channel.hpp

```diff
@@ -0,0 +1,304 @@
+#ifndef DICE_TEMPLATELIBRARY_CHANNEL_HPP
+#define DICE_TEMPLATELIBRARY_CHANNEL_HPP
+
+#include <atomic>
+#include <cassert>
+#include <condition_variable>
+#include <cstddef>
+#include <deque>
+#include <mutex>
+#include <optional>
+
+namespace dice::template_library {
+
+	/**
+     * A single producer, single consumer channel/queue
+     * @note this can technically be used as a multi producer, single consumer queue, but care must be taken
+     *      of when exactly close() is called
+     * @warning close() must be called once the producing thread is done, otherwise the reading thread will hang indefinitely or SEGFAULT
```

Comment:
I thought about it again, and yeah the SEGFAULT would be less the result of not calling close() and more because of improper lifetime management of threads. I.e. not joining all threads before the destructor. I will remove the note